pkg/services/ngalert/schedule/recording_rule.go

```diff
@@ -212,3 +240,17 @@ func (r *recordingRule) evaluationDoneTestHook(ev *Evaluation) {
 
 	r.evalAppliedHook(ev.rule.GetKey(), ev.scheduledAt)
 }
+
+func (r *recordingRule) frameRef(refID string, resp *backend.QueryDataResponse) (data.Frames, error) {
+	if len(resp.Responses) == 0 {
+		return nil, fmt.Errorf("no responses returned from rule evaluation")
+	}
+
+	for ref, resp := range resp.Responses {
+		if ref == refID {
+			return resp.DeepCopy().Frames, nil
```

Comment:
The response object can be absolutely enormous in certain cases - megabytes even, it contains every single result of every single expression node. So, number of dimensions * number of expression nodes. Is there a way we can avoid a deep copy?

Reply:
Whoops that shouldn't have made it in, was using this for debugging.